### PR TITLE
Train refactoring

### PR DESCRIPTION
**Pull request by YuriiLaba**

### CODE REVIEW

File: services/trainer/trainer.py [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+                acc_val, acc_avg = self.train_avg_meter()
+                self.run["train/loss"].append(acc_avg)
+
+            report_gpu() # TODO: it's interesting do we really need it
+
+            if batch_count % self.config.getint('MODEL_TUNING', 'num_batch_to_eval') == 0 and batch_count != 0:
+                if self.evaluate_epoch(epoch, batch_count):
+                    return # reach early stopping rounds
+    
+    def train(self):
```
Comment (Mudryi) [inline services/trainer/trainer.py:245]: need to add in this function run.stop() if neptune True

File: services/trainer/trainer.py [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+    def _save_model(self, model, path_to_save_model):
+        try:
+            if isinstance(model, torch.nn.DataParallel):
+                model.module.save_pretrained(path_to_save_model, from_pt=True)
+            else:
+                model.save_pretrained(path_to_save_model, from_pt=True)
+        except Exception as e:
+            print(f'model not saved, error = {e}')
+
+    def train_step(self, batch):
```
Comment (Mudryi) [inline services/trainer/trainer.py:176]: rename to calculate_loss

File: services/trainer/trainer.py [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+        self.train_avg_meter = AverageMeter("train_loss")
+        self.max_wsd_acc = 0
+        self.rounds_count = 0
+
+        # TODO: i'd like to have better config parsing. It takes too much space 
+        self.optim = torch.optim.Adam(self.model.parameters(), lr=self.config.getfloat('MODEL_TUNING', 'learning_rate'))
+
+        if self.apply_warmup:
+            total_steps = int(len(self.train_dataset) / config.getint('MODEL_TUNING', 'batch_size'))
+            warmup_steps = int(0.1 * total_steps)
```
Comment (Mudryi) [inline services/trainer/trainer.py:106]: we can move 0.1 to config as warmup ration

File: services/trainer/trainer.py [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+
+    def train_epoch(self, epoch):
+        train_bar = tqdm(self.train_loader, leave=True, desc=f'Train epoch: {epoch}')
+
+        for batch_count, batch in enumerate(train_bar):
+            self.model.train()
+            loss = self.train_step(batch)
+            loss.backward()
+
+            self.train_avg_meter.update(loss.item(), self.config.getint('MODEL_TUNING', 'batch_size'))  # TODO: .detach().cpu()?
```
Comment (Mudryi) [inline services/trainer/trainer.py:227]: move. this line after self.optim.zero_grad()

File: services/trainer/trainer.py [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+            if self.rounds_count == self.config.getint('MODEL_TUNING', 'early_stopping'):
+                print(f'Early stopping, model not improve WSD for {self.config.getint("MODEL_TUNING", "early_stopping")}')
+                return True
+        return False
+
+    def train_epoch(self, epoch):
+        train_bar = tqdm(self.train_loader, leave=True, desc=f'Train epoch: {epoch}')
+
+        for batch_count, batch in enumerate(train_bar):
+            self.model.train()
```
Comment (Mudryi) [inline services/trainer/trainer.py:223]: move self.model.train() after eval to run in only if model was on eval mode

**Review by Mudryi — APPROVED** [review]
